packages/rest/src/rest.ts

```diff
@@ -76,15 +77,84 @@ const discriminatorKey = Symbol("discriminator");
 
 const discriminatorDecorator = createDecoratorDefinition({
   name: "@discriminator",
-  target: "Model",
+  target: ["Model", "Union"],
   args: [{ kind: "String" }],
 } as const);
 
-export function $discriminator(context: DecoratorContext, entity: ModelType, propertyName: string) {
+export function $discriminator(
+  context: DecoratorContext,
+  entity: ModelType | UnionType,
+  propertyName: string
+) {
   if (!discriminatorDecorator.validate(context, entity, [propertyName])) {
     return;
   }
 
+  let hasErrors = false;
+  if (entity.kind === "Union") {
+    // we can validate discriminator up front for unions. Models are validated
+    // in emitters.
+    for (const variant of entity.variants.values()) {
+      if (typeof variant.name === "symbol") {
+        // likely not possible to hit this without applying the decorator programmatically
+        reportDiagnostic(context.program, {
```

Comment:
should this then be a compilerAssert?

Reply:
I guess maybe not, feel like its not a great experience either way. At least here it wouldn't crash compilation if that mistake was made and not caught

Reply:
I went back and forth. I don't have a strong rationale either way tbh.